### ISSUE TITLE
Implement Zoom In: Neuron Feature Visualization

### ISSUE THREAD

**Issue by mly-johndpope** [johndpope/latent-tools#2]
## Overview

Implement neuron-level feature visualization similar to [Distill: Zoom In](https://distill.pub/2020/circuits/zoom-in/). This complements the weight visualization (#1) by showing what individual neurons actually detect in images.

## Goal

Visualize what individual neurons/channels respond to by:
1. Generating synthetic images that maximally activate specific neurons
2. Finding real image patches that strongly activate neurons
3. Creating activation atlases to understand neuron behavior

### Feature Visualization
- **Activation Maximization**: Generate images that maximize a specific neuron's activation
- **Diversity**: Show multiple examples of what activates a neuron
- **Interactive**: Allow users to select which neuron/channel to visualize

### Activation Atlas
- Show spatial activation patterns across an image
- Visualize which parts of an image activate specific neurons
- Compare activations across different neurons

### Dataset Examples
- Find and display real image patches that strongly activate a neuron
- Rank and sort by activation strength
- Show diverse examples

## Reference Implementation

The Distill Zoom In article shows:
1. **Feature Visualization**: Synthetic images showing what neurons detect
   - Edge detectors, texture detectors, pattern detectors
   - Multiple diverse examples per neuron
   - Clear, interpretable visualizations

2. **Activation Atlas**: Spatial maps of neuron activations
   - Heatmaps showing where neurons fire
   - Comparison across layers

3. **Dataset Examples**: Real image crops
   - Top-activating patches from real images
   - Helps validate synthetic visualizations

## Proposed ComfyUI Integration

Create new nodes:

### LTFeatureVisualization
Generate synthetic images that maximize neuron activation.

**Inputs:**
- Model/checkpoint
- Layer name
- Channel/neuron index
- Number of examples to generate
- Optimization steps
- Image size

**Outputs:**
- Generated feature visualizations
- Activation values
- HTML preview

### LTActivationAtlas
Show spatial activation patterns across an input image.

**Inputs:**
- Model
- Layer name
- Input image (or latent)
- Channels to visualize

**Outputs:**
- Activation heatmaps
- Channel-wise activation maps
- HTML preview with overlays

### LTFindActivatingPatches
Find image patches that strongly activate a neuron.

**Inputs:**
- Model
- Layer name
- Channel index
- Image dataset/folder
- Number of top patches

**Outputs:**
- Top-N image patches
- Activation scores
- HTML gallery

### 1. Feature Visualization (Activation Maximization)

Use gradient ascent to generate images:
```python
# Start with random noise
image = torch.randn(1, 3, 224, 224, requires_grad=True)

for step in range(num_steps):
    # Forward pass
    activations = get_layer_activation(model, image, layer_name)
    
    # Maximize target neuron activation
    loss = -activations[0, channel_idx].mean()
    
    # Backward pass
    loss.backward()
    
    # Update image
    image.data -= learning_rate * image.grad
    image.grad.zero_()
    
    # Apply regularization (total variation, etc.)
    image.data = apply_regularization(image.data)
```

### 2. Activation Atlas

Hook into model layers to capture activations:
```python
activations = {}

def hook_fn(module, input, output):
    activations[layer_name] = output

# Register hook
handle = layer.register_forward_hook(hook_fn)

# Forward pass with input image
model(input_image)

# Visualize activations[layer_name]
```

### 3. Dataset Examples

For each image in dataset:
- Forward pass through model
- Extract activation at target layer/channel
- Compute activation strength
- Keep top-K patches

## Regularization Techniques

To get clean, interpretable visualizations:
1. **Total Variation**: Encourage spatial smoothness
2. **L2 Decay**: Prevent extreme pixel values
3. **Jitter**: Random translations during optimization
4. **Random Scaling**: Multi-scale features
5. **Rotation**: Rotation invariance
6. **Color Decorrelation**: Use natural color statistics

## Integration with Weight Visualization

Combine with existing weight visualization:
- Weights show *what the filter is*
- Feature viz shows *what the filter detects*
- Together: complete understanding of a neuron

## Example Use Cases

1. **Understanding SDXL layers**: See what textures/patterns each layer detects
2. **Model debugging**: Identify what features the model learned
3. **Creative exploration**: Discover interesting neuron behaviors
4. **Model comparison**: Compare features across different checkpoints

## References

- [Distill: Zoom In](https://distill.pub/2020/circuits/zoom-in/)
- [Distill: Feature Visualization](https://distill.pub/2017/feature-visualization/)
- [Activation Atlas](https://distill.pub/2019/activation-atlas/)
- [Lucid library](https://github.com/tensorflow/lucid) (TensorFlow implementation)

## Benefits

- Understand what SDXL has learned without training
- Debug unexpected model behavior
- Educational tool for understanding deep learning
- Creative tool for discovering novel visual patterns
- Complements weight visualization for complete understanding

## Implementation Priority

1. **Phase 1**: Basic feature visualization (activation maximization)
2. **Phase 2**: Activation atlas (spatial maps)
3. **Phase 3**: Dataset examples (real image patches)
4. **Phase 4**: Interactive exploration and refinement
